Add automatic whitelist generator for vulture

RJ722 opened the pull request:
Closes: #563

AbdealiJK commented on _projects/automate_vulture_whitelist.md:
```diff
@@ -0,0 +1,77 @@
+collaborating_projects:
+  - coala
+  - vulture
+desc: "Automatically create whitelists and detect falsely positive results for vulture."
```
grammar: `false positives`

AbdealiJK commented on _projects/automate_vulture_whitelist.md:
```diff
@@ -0,0 +1,77 @@
+* The script for creating a whitelist from sip files should be implemented.
+* It should reside in it's own repository, with CI enabled for testing on both
+  the GNU/Linux and Windows platforms.
+* It should be gracefully documented.
```
gracefully documented :) :+1:

AbdealiJK commented on _projects/automate_vulture_whitelist.md:
```diff
@@ -0,0 +1,77 @@
+* It should be gracefully documented.
+* Use this script to create a whitelist for
+  [qutebrowser](https://github.com/qutebrowser/qutebrowser).
+* Increase vulture's code coverage to 100%.
```
Sure you want to commit to 100% ?
@jendrikseipp mentioned this may involve a very large change to the current architecture. Maybe just mention "look at methods to increase coverage for vulture from it's already excellent 95%+ coverage"?

AbdealiJK commented on _projects/automate_vulture_whitelist.md:
```diff
@@ -0,0 +1,77 @@
+* [Follow up milestone](https://gitlab.com/coala/GSoC/GSoC-2018/milestones/29)
+
+##### CODING PHASE 2
+*  `--make-whitelist`: Vulture should be able to find a list of false positives
```
Minor: There's an extra space at the start here

AbdealiJK commented:
ack for the overall PR  - but As we spoke earlier - the blog seems to be down.
The Blog seems to be down due to technical issues but that should be up in a day or two

@jayvdb - can we merge this or wait for the blog to get back up?

RJ722 commented:
> The Blog seems to be down due to technical issues but that should be up in a day or two

There's an ongoing PR already: #572

jayvdb commented on _projects/automate_vulture_whitelist.md:
```diff
@@ -0,0 +1,78 @@
+
+##### CODING PHASE 1
+* The script for creating a whitelist from sip files should be implemented.
+* It should reside in it's own repository, with CI enabled for testing on both
```
is this a coala org repo, or is vulture going to become an org?  or under your own user space?

Presumably these whitelists could also be used by coala's python bears, and other python tools, which are currently unaware of voodoo provided by frameworks and C wrappers.

RJ722 replied:
> is this a coala org repo, or is vulture going to become an org? or under your own user space?

I'd very much like input from @AbdealiJK and @jendrikseipp.

> Presumably these whitelists could also be used by coala's python bears, and other python tools, which are currently unaware of voodoo provided by frameworks and C wrappers.

I cannot see a simple use case where other coala bears or python tools can use these whitelists. Do you have any ideas in specific @jayvdb?

The-Compiler replied:
The whitelist is probably useful for any project using PyQt (or PySide2), but not for other projects.

jendrikseipp replied:
I'd call the repo RJ722/vulture-whitelist-generators for now. I think it'll become clearer where to put this later.

RJ722 replied:
> I'd call the repo RJ722/vulture-whitelist-generators for now. I think it'll become clearer where to put this later.

Hehe, Thanks for the honor @jendrikseipp!

jayvdb replied:
> I cannot see a simple use case where other coala bears or python tools

All of them could use it, for the same reasons as vulture.  Maybe not simple.

e.g. see our pyflakes project cEP.

jayvdb replied:
> Whitelists are just Python files containing a reference to an object which was reported as unused by vulture

This synopsis is precisely why the problem should not be though of in terms of only vulture.

The work being done is not identifying references to an object or method which was reported as unused by vulture.

The work being done is identifying references to ast elements that not unused, *but* no Python AST-only based algorithms can detect without inspecting external definition files and/or coverage data, or at least these means of gathering the data is 'better' than trying to obtain it from Python AST of dependencies.

It should be a generic tool (library) which provides the information which was determined by using sip/shiboken/coverage, and then a vulture specific use of that information which pushes that information into the whitelist format which vulture wants.



pyflakes has a whitelist mechanism, the [`PYFLAKES_BUILTINS`](https://github.com/PyCQA/pyflakes/blob/a13dbe6/pyflakes/checker.py#L478) envvar, which could be populated by data obtained from your tools, especially the information obtained from coverage.xml.

This same library could also be utilised by a new bear using [pyflakes AST](https://github.com/coala/cEPs/blob/master/cEP-0025.md), which is @MacBox7's project.

Bears written using the antlr AST will likely have the same problems.

MacBox7 replied:
in order to achieve this we would be needing an inverse mapping i.e given the line and column provide us with `pyflakes node`. We could then get a list of unused code from some tools (Vulture or pyflakes itself) and cross check it with coverage report to generate whitelist.

jayvdb replied:
this project would only provide a list of symbols found in scopes *not* part of the pyflakes ast analysis - see project for where it plans to get them from. It isnt necessary to know line or col : the primary value is in knowing about extra names which are valid.

These symbol lists are only useful to augment the inbuilt data. See pyflakes envvar mentioned in my last comment. Specifically to prevent false positives being report.

building an unused code detector bear is not a useful use case as that would be reinventing the vulture wheel.

MacBox7 replied:
I understand this. But the coverage report returned by [coverage.py](https://coverage.readthedocs.io/en/coverage-4.5.1/) only returns lineno and not the list of symbols. Thus an inverse mapping is required to return nodes corresponding to the line no. These symbols can then be supplied to envvar.

RJ722 replied:
The workflow would be something like this:
- run Vulture on the project
- whatever functions or methods vulture reports as unused, check whether or not that particular line number(s) are hit or not in the `coverage.xml`'s output.

An inverse mapping would be required if you are going the opposite way, from `coverage`'s output to the `ast`.

jayvdb commented on _projects/automate_vulture_whitelist.md:
```diff
@@ -0,0 +1,78 @@
+* [Follow up milestone](https://gitlab.com/coala/GSoC/GSoC-2018/milestones/28)
+
+##### CODING PHASE 1
+* The script for creating a whitelist from sip files should be implemented.
```
`script` implies junky code.

You are creating a new repo, presumably a new package, and all code in it will be lovely.
And it will have a command line entry point.

Below you have this "script to create a whitelist for qutebrowser" , which is where the lower quality comes in -- in milestone 1 you are expecting that the very high quality code will be only guaranteed to support qutebrowser's needs.

I expect that this means that in milestone 1 you are supplying a PR to qutebrowser which replaces their whitelist generator with your own.

In the following milestones more use cases are added, and that will expand the scope of the codebase, dealing with additional scenarios not needed by qutebrowser.

The-Compiler replied:
> I expect that this means that in milestone 1 you are supplying a PR to qutebrowser which replaces their whitelist generator with your own.

That [whitelist generator](https://github.com/qutebrowser/qutebrowser/blob/master/scripts/dev/run_vulture.py) would still need to stay, because it has many things in it which are qutebrowser specific. What it'd replace is the current [horrible heuristic](https://github.com/qutebrowser/qutebrowser/blob/v1.3.0/scripts/dev/run_vulture.py#L136) which ignores any functions which have a capital letter in them (because Qt/PyQt use `camelCase` for methods).

jayvdb replied:
I guess that means that the whitelist generator would need to also provide a API for the qutebrowser whitelist generator to use.

RJ722 replied:
No, we wouldn't need an API. The whitelist needs to be manually generated, saved and committed into the project. Even better would be when we start shipping it with vulture - That way whitelist would be automatically loaded whenever user imports Qt.

jayvdb commented on _projects/automate_vulture_whitelist.md:
```diff
@@ -0,0 +1,78 @@
+* It should be gracefully documented.
+* Use this script to create a whitelist for
+  [qutebrowser](https://github.com/qutebrowser/qutebrowser).
+* Look at methods to increase coverage for vulture from it's already excellent
```
"Look at methods" is not a deliverable.
Try:

Raise the coverage of vulture to 100% with limited use of `#pragma: no cover` where agreed upon by the vulture maintainers.

RJ722 replied:
I don't think that @jendrikseipp agrees. He feels like code readability decreases by littering the code base with comments.

cc @jendrikseipp

jendrikseipp replied:
That's true. It's no problem if code coverage is less than 100% in my opinion. But trying to raise it is a good idea.

RJ722 replied:
If we want a deliverable, we may say that: "Maximise code coverage for vulture without littering it with comments." Do you agree @jayvdb?

jayvdb replied:
If using `#pragma: no cover` to reach 100%, it shouldnt be a style concern as they should be used exceedingly sparingly, in the very difficult to reach parts of the codebase, and having a specific machine readable comment in those regions helps people reading the code know that those lines are *special*, as they are not covered, and are difficult to reach in the test suite -- i.e. dragons be there; dont hack this or merge related changes before you've had a strong coffee.

coalib (coala core) only has 21 of them, and we have PRs to remove more of them.  coverage also allows you to use a different comment text, if you dont like `#pragma`.

Anyway .. it isnt a coala project, so the decision here needs to be made by the maintainer .   "Maximise code coverage for vulture" is much better than "Look at .." :+1:

jayvdb commented on _projects/automate_vulture_whitelist.md:
```diff
@@ -0,0 +1,78 @@
+* [Follow up milestone](https://gitlab.com/coala/GSoC/GSoC-2018/milestones/29)
+
+##### CODING PHASE 2
+* `--make-whitelist`: Vulture should be able to find a list of false positives
```
have you choosen a library for parsing `coverage.xml`?

RJ722 replied:
I haven't decided yet, but I'm thinking of using `lxml`.

The-Compiler replied:
Definitely a decent and powerful library! If you can wrap your head around its documentation (I remember that taking me a while back when I first used it - but don't let that discourage you, your mileage may vary), it's definitely a good choice.

"The Hitchhiker's Guide to Python" also has some [simpler alternatives](http://docs.python-guide.org/en/latest/scenarios/xml/), but I'm not sure if they'll be powerful enough for this use-case.

RJ722 replied:
Thanks @The-Compiler! 

I thought of using `lxml` to maintain consistency with the script you wrote [here](). 

>  "The Hitchhiker's Guide to Python" also has some simpler alternatives, but I'm not sure if they'll be powerful enough for this use-case.

jayvdb commented on _projects/automate_vulture_whitelist.md:
```diff
@@ -0,0 +1,78 @@
+##### CODING PHASE 2
+* `--make-whitelist`: Vulture should be able to find a list of false positives
+  (for functions and methods) from `coverage.xml`.
+* Maintain code coverage, robust tests should be written.
```
-> Maintain 100% test code coverage established in coding phase 1.

jayvdb replied:
This comment depends on the outcome of this discussion https://github.com/coala/projects/pull/571#discussion_r187789504 ; maybe 100% isnt desirable.

jayvdb commented on _projects/automate_vulture_whitelist.md:
```diff
@@ -0,0 +1,78 @@
+##### CODING PHASE 3
+* Implement whitelist generator for
+  [`shiboken`](https://pypi.org/project/Shiboken/).
+* Add tests and documentation for the generator.
```
Split into two items:

- Maintain 100% test code coverage.
- Update documentation for the whitelist generator

jayvdb commented on _projects/automate_vulture_whitelist.md:
```diff
@@ -0,0 +1,78 @@
+##### CODING PHASE 1
+* The script for creating a whitelist from sip files should be implemented.
+* It should reside in it's own repository, with CI enabled for testing on both
+  the GNU/Linux and Windows platforms.
```
... and tests providing 100% code coverage with limited use of `#pragma: no cover`

jayvdb commented on _projects/automate_vulture_whitelist.md:
```diff
@@ -0,0 +1,78 @@
+
+#### IDEAS
+
+* Create a script which takes sip files as input and spits a whitelist with all
```
It would be useful to also develop a SIP Bear (and .sbf?) as a background task during your project - they can just be simple pass/fail based on `sipconfig` parser failing, but ideally becomes a little more functional over the course of your project.

RJ722 replied:
Let me quote your email here because I guess, the mailing list isn't public:

> I doubt that it'd be useful - SIP is rarely used outside of PyQt.
>
> As for PyQt, it's pretty much a one-man show which doesn't really
accommodate external contributors unfortunately - for example, no public
repository exists.
>
> Also, they're mostly auto-generated (in a half-automatic way) as far as
I'm aware - the .sip format is pretty much a C++ header file (.h) with
additional annotations, so projects have some way of keeping those two
in sync.
>
> I think in case of PyQt, the author has some kind of IDE which parses .h
files, asks him for the annotations for any new functions (when a new Qt
version gets released) and then spits out the .sip files.

The-Compiler commented on _projects/automate_vulture_whitelist.md:
```diff
@@ -0,0 +1,78 @@
+  the GNU/Linux and Windows platforms.
+* It should be gracefully documented.
+* Use this script to create a whitelist for
+  [qutebrowser](https://github.com/qutebrowser/qutebrowser).
```
This again implies the whitelist would be qutebrowser specific - but it isn't, it's a whitelist for e.g. PyQt5 which can be used in any project using PyQt5.

RJ722 replied:
Yes, this sentence is also wrong. I'll change it to:

* A whitelist for PyQt should be generated which should then be shipped with vulture by default.

The-Compiler commented on _projects/automate_vulture_whitelist.md:
```diff
@@ -0,0 +1,78 @@
+* Add documentation, how-to, examples and use cases in
+  [README](https://github.com/jendrikseipp/vulture/tree/master/README.rst).
+* Research and planning about how to generate whitelists for
+  [`shiboken`](https://pypi.org/project/Shiboken/)
```
That seems like the old Shiboken used in PySide (not PySide2), seeing that the last update was in 2014. It's a bit of a mess to be honest, but I think the current correct location is [this folder](http://code.qt.io/cgit/pyside/pyside-setup.git/tree/sources/shiboken2) in the `pyside-setup` repository.

RJ722 commented on _projects/automate_vulture_whitelist.md:
```diff
@@ -0,0 +1,78 @@
+
+The aim of this project is to automatically create whitelists (to be utilized
+by [vulture](https://github.com/jendrikseipp/vulture)) for projects using sip
+files to communicate between C++ level API and Python programs, most
```
Yes, that definitely sounds more appealing. Thanks @The-Compiler!

RJ722 commented on _projects/automate_vulture_whitelist.md:
```diff
@@ -0,0 +1,78 @@
+The aim of this project is to automatically create whitelists (to be utilized
+by [vulture](https://github.com/jendrikseipp/vulture)) for projects using sip
+files to communicate between C++ level API and Python programs, most
+significant of them being PyQt projects.
```
Let me rephrase it:

The aim of this project is to automatically create whitelists (to be utilized by [vulture](https://github.com/jendrikseipp/vulture)) for projects using sip files to generate Python bindings for C++ code, most significant of them being `PyQt`. Once implemented, a whitelist  generated through this generator for `PyQt` could be universally utilised across all `PyQt` projects.

Does it sound any better?

The-Compiler replied:
Sounds good to me!

RJ722 commented:
I've intentionally added another commit so that the context for this discussion is preserved. I'll squash the commits after everyone is happy with the changes.

jayvdb commented on _projects/automate_vulture_whitelist.md:
```diff
@@ -0,0 +1,82 @@
+* It should reside in it's own repository, with CI enabled for testing on both
+  the GNU/Linux and Windows platforms.
+* It should be gracefully documented.
+* A whitelist for PyQt should be generated to be then shipped with vulture by
```
@jendrikseipp agrees with bundling the PyQt whitelist into the vulture releases?

That doesnt sound very scalable, as then vulture would be asked to ship other whitelists.  Generation of whitelists feels like a task that vulture users should do, and possibly store in their repo and regenerate periodically when their dependencies change.

RJ722 replied:
> @jendrikseipp agrees with bundling the PyQt whitelist into the vulture releases?

I know that he he didn't like this idea initially, but since then, there have been some major advancements in the way vulture handles whitelists. Now, we've whitelists for every library for which we know that vulture raises certain false positives (https://github.com/jendrikseipp/vulture/tree/master/vulture/whitelists). Using this support infrastructure for whitelisting, it would be a lot nicer to ship whitelist with vulture, wouldn't it @jendrikseipp?

> That doesnt sound very scalable, as then vulture would be asked to ship other whitelists. 

Yes, it does requires some efforts for maintaining those, but I guess, owing to how much user friendly it makes vulture, it's worth it. And also, it was also one of my task to create whitelists for popular libraries during last year's GSoC, remember?

> Generation of whitelists feels like a task that vulture users should do, and possibly store in their repo and regenerate periodically when their dependencies change.

Since not all whitelists need to be dynamically generated, it wouldn't be very user friendly to let users do the mundane work of creating a whitelist and including it with vulture.

jayvdb commented on _projects/automate_vulture_whitelist.md:
```diff
@@ -0,0 +1,82 @@
+  default, thus helping PyQt projects (like
+  [qutebrowser](https://github.com/qutebrowser/qutebrowser)) to deal with false
+  positives.
+* Maximise code coverage of vulture.
```
this isnt what I said in last review.

Maximise is a process, not a goal.  100% is a goal.  You can reach 100% without hitting every line by using `#pragma: no cover` so there is no reason why you should not be able to achieve this goal.

jayvdb commented on _projects/automate_vulture_whitelist.md:
```diff
@@ -0,0 +1,82 @@
+
+##### CODING PHASE 1
+* The whitelist generator should be completely implemented.
+* It should reside in it's own repository, with CI enabled for testing on both
```
this repo/project needs to maintain 100% test coverage.

very likely this will be part of coala , if not bears, it would need to be part of the `coala-quickstart` program (a project for 2019!) to generate the whitelist for VultureBear and maybe other bears also.

jayvdb replied:
This comment has been addressed below.

jayvdb commented on _projects/automate_vulture_whitelist.md:
```diff
@@ -0,0 +1,82 @@
+##### CODING PHASE 1
+* The whitelist generator should be completely implemented.
+* It should reside in it's own repository, with CI enabled for testing on both
+  the GNU/Linux and Windows platforms.
```
CI should also ensure ongoing support for 2.7 & py34+ for compatibility with vulture

jayvdb commented:
Your commit message doesnt refer to https://github.com/coala/projects/issues/563

What is going on with the rj722.tech link?  a gitmate bug?
Maybe add a commit to this PR to bypass the bug, like https://github.com/coala/projects/commit/4a482dd96083b0cd7c0ebcc32ee55fa0b0c8ef76 until your website is working :/

RJ722 commented:
> Your commit message doesnt refer to #563

Done!

> What is going on with the rj722.tech link? a gitmate bug?
Maybe add a commit to this PR to bypass the bug, like 4a482dd until your website is working :/

Actually, I don't plan to use rj722.tech for now. #572 has been merged, so I guess it isn't a problem anymore.

jayvdb commented on _projects/automate_vulture_whitelist.md:
```diff
@@ -0,0 +1,86 @@
+status: in_progress
+initiatives:
+  - GSoC
+issues: []
```
Please fill this with relevant issues in vulture & coala repos

jayvdb commented on _projects/automate_vulture_whitelist.md:
```diff
@@ -0,0 +1,86 @@
+
+The main aim of this project is to automatically create whitelists (to be
+utilized by vulture) for projects projects using
+[sip](https://www.riverbankcomputing.com/software/sip/intro) and
```
please check you are using backticks everywhere that is appropriate.

jayvdb commented on _projects/automate_vulture_whitelist.md:
```diff
@@ -0,0 +1,86 @@
+  [qutebrowser](https://github.com/qutebrowser/qutebrowser)) to deal with false
+  positives.
+* Maximise code coverage of vulture.
+* [Follow up milestone](https://gitlab.com/coala/GSoC/GSoC-2018/milestones/29)
```
please do not include this type of bullet point about milestones in https://gitlab.com/coala/GSoC/GSoC-2018/

The project page should have been published before https://gitlab.com/coala/GSoC/GSoC-2018/ was created, and before the gsoc slots were selected.  You need to pretend those milestones dont exist yet.

If there is anything in those milestones which should be added to the project, they need to be added as bullet points here.

jayvdb commented on _projects/automate_vulture_whitelist.md:
```diff
@@ -0,0 +1,85 @@
+##### COMMUNITY BONDING
+* Discussion on whether to implement a new Tracer or to directly consume `xml`
+  output from `coverage.py`.
+* Get familiar with `coverage.xm`.
```
`coverage.xml`

jayvdb commented on _projects/automate_vulture_whitelist.md:
```diff
@@ -0,0 +1,85 @@
+#### MILESTONES
+
+##### COMMUNITY BONDING
+* Discussion on whether to implement a new Tracer or to directly consume `xml`
```
`xml` -> XML

jayvdb commented on _projects/automate_vulture_whitelist.md:
```diff
@@ -0,0 +1,85 @@
+  will reduce overhead for PyQt projects which use vulture.
+* Allow passing coverage.xml to vulture via `--make-whitelist` and let it find
+  the list of false-positive functions. (Add Documentation and tests as well)
+* Write a whitelist generator for shiboken, the binding generator used for
```
trailing full stop missing

jayvdb replied:
(fixed)

jayvdb commented on _projects/automate_vulture_whitelist.md:
```diff
@@ -0,0 +1,85 @@
+* Write tests, achieve 100% code coverage.
+* It should be gracefully documented.
+* A whitelist for PyQt should be generated to be then shipped with vulture by
+  default, thus helping PyQt projects (like
```
round brackets not particularly useful/necessary here.

jayvdb commented on _projects/automate_vulture_whitelist.md:
```diff
@@ -0,0 +1,85 @@
+  [`shiboken`](http://code.qt.io/cgit/pyside/pyside-setup.git/tree/sources/shiboken2).
+* Add new tests, sustain code coverage.
+* Update documentation for the whitelist generator.
+
```
remove trailing blank line.

RJ722 commented:
I've addressed your comments @jayvdb and sorry for trivial mistakes, I've started using vim "full time" recently and that's why these slipped by under my nose so easily. I hope it will get better with time. Thanks! :-)

jayvdb commented on _projects/automate_vulture_whitelist.md:
```diff
@@ -0,0 +1,84 @@
+  [TODO](https://github.com/jendrikseipp/vulture/tree/master/TODO.rst).
+
+##### CODING PHASE 1
+* The whitelist generator should be completely implemented.
```
As mentioned at https://github.com/coala/projects/pull/571#pullrequestreview-119643407 , it would be good if it was designed foremost a re-usable library which did the hard work with a decent API, and then a CLI interface which did the 'vulture' specific whitelist generation.

jayvdb replied:
I see nothing in the project brief about this, and no direct reply here.  Please reply directly on the PR diff to keep the conversion together.  https://github.com/coala/projects/pull/571#issuecomment-392283437 seems to be a reply, where you tell me 'no'.
Why do you want to write a tool which can not be used by other people?

RJ722 replied:
Hehe @jayvdb! I didn't say that no one would use it - I'm just saying that it should be user ended rather than being developed with an API.

Here, in this case, we are trying to develop a CLI app to create whitelists for vulture (via looking up for virtual functions in the XML output) and parsing other information from XML and presenting it is not something to look up to.

jayvdb replied:
Only the output format is vulture specific.  That is the easy part.  The difficult part is all of the parsing, and that is usable by coala, and you are refusing to build it with functions that coala can use.

jayvdb replied:
Im not sure what the problem is with exposing an API, but assuming the code is half decently designed, we can add it later, so closing this off for now.

RJ722 commented:
>  split this into two items that are more specific; c.f. #571 (review)

Done. I've also added other requirements.

> Also maybe it would be useful to put some reports of PyQt and PySide2 projects into https://github.com/jendrikseipp/vulture-analysis , capturing before and after for this project.

That is a great idea.

> The work being done is identifying references to ast elements that not unused, but no Python AST-only based algorithms can detect without inspecting external definition files.

Actually, whitelists don't work because we find a reference to the original (unused) node in the ast. It works because Vulture indexes every node in a set with it's name and therefore when it finds out that the names of the unused node and the whitelist node are same, it "thinks" that the dead code is somehow being used here and thereby does not report it as unused.

>  It should be a generic tool (library) which provides the information which was determined by using sip/shiboken/coverage, and then a vulture specific use of that information which pushes that information into the whitelist format which vulture wants.

> As mentioned at #571 (review) , it would be good if it was designed foremost a re-usable library which did the hard work with a decent API, and then a CLI interface which did the 'vulture' specific whitelist generation.

The work we are doing for coverage involves looking up a particular line number in the output from coverage's XML for whether or not the `hits` switch is on. IMHO this is nothing fancy that needs to be generalised.

For PyQt, we are trying to index the XML files in order to find all the virtual functions defined in the `sip` files. I'm not sure about `shiboken`, but I guess we would be doing something similar. I can't see a simple use case of this functionality as a library.

RJ722 commented:
Hey,

#587 has been merged, can I get a green flag on this now 😅 ?

cc @jayvdb  @AbdealiJK @jendrikseipp @The-Compiler

jayvdb commented:
fyi ongoing discussion at https://github.com/coala/projects/pull/571#discussion_r190448402

jayvdb commented:
ack https://github.com/coala/projects/pull/571/commits/06d27c197cc8e37c5b60ecddd61be0c4c497f8ff

jayvdb commented:
@gitmate-bot rebase

jayvdb commented:
@gitmate-bot ff
